### PR TITLE
feat(runtime): support RSA-PSS, RSASSA-PKCS1-v1_5 & ECDSA for `SubtleCrypto#sign` & `SubtleCrypto#verify`

### LINKED ISSUE #407: ECDSA for `SubtleCrypto#sign`  & `SubtleCrypto#verify`
We should support the ECDSA algorithm for `SubtleCrypto#sign` and `SubtleCrypto#verify` as required by [W3C](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto#supported_algorithms) and [WinterCG](https://proposal-common-min-api.deno.dev/). Right now, only HMAC [is supported](https://docs.lagon.app/runtime-apis#cryptosubtle).

### LINKED ISSUE #408: RSA-PSS for `SubtleCrypto#sign`  & `SubtleCrypto#verify`
We should support the RSA-PSS algorithm for `SubtleCrypto#sign` and `SubtleCrypto#verify` as required by [W3C](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto#supported_algorithms) and [WinterCG](https://proposal-common-min-api.deno.dev/). Right now, only HMAC [is supported](https://docs.lagon.app/runtime-apis#cryptosubtle).

### PR DESCRIPTION
**Pull request by akitaSummer**
Hello

I have implemented sign and verify functions for RSASSA-PKCS1-v1_5, RSA-PSS, and ECDSA, but only the test for ECDSA is passing. I suspect it may be due to the large size of Rust's vec<u8>, causing errors when converting to V8's ArrayBuffer. I suggest trying to find a solution to this issue, or perhaps using serde_v8's ZeroCopyBuf to refactor the project.
I have refactored the generateKey method to support all algorithms required by W3C, but I have not handled the extractable and keyUsages parameters. I hope you can address these when you have time.
I reviewed your importKey method and found that it does not comply with W3C standards. I suggest that you consider modifying it when you have the time.

Closes #408
Closes #407

### CODE REVIEW

**Comment by QuiiBz** [comment]
Thanks for the PR!

> I suspect it may be due to the large size of Rust's vec
I'll have a deeper look today, but that seems strange.

`importKey` was definitely wrong, I actually made https://github.com/lagonapp/lagon/pull/860 a few hours ago but looks like it's now timeouting the CI. `keyUsages` & `extractable` will be supported later.

**Comment by QuiiBz** [comment]
I pushed some fix (most notably https://github.com/lagonapp/lagon/pull/862), we now only have a single test failing (due to a timeout): crypto_rsa_pss_sign_verify

Screenshot 2023-05-14 at 10 25 20

File: crates/runtime_crypto/Cargo.toml [hunk 1]
```diff
@@ -17,5 +17,8 @@ aes = "0.8.2"
 aes-gcm = "0.10.1"
 cbc = { version = "0.1.2", features = ["std"] }
 ring = { version = "0.16.20", features = ["std"] }
+once_cell = "1.10.0"
+num-traits = "0.2.14"
+rsa = { version = "=0.7.0-pre", default-features = false, features = ["std"] }
```
Comment (QuiiBz) [inline crates/runtime_crypto/Cargo.toml:22]: Could we use the latest version of these crates?

```suggestion
once_cell = "1.17.1"
num-traits = "0.2.15"
rsa = { version = "=0.9.2", default-features = false, features = ["std"] }
```

File: crates/runtime/tests/crypto.rs [hunk 2]
```diff
@@ -548,3 +567,89 @@ async fn crypto_derive_key() {
...
+    send(Request::default());
+
+    assert_eq!(
+        receiver.recv_async().await.unwrap().as_response(),
+        Response::from("true")
+    );
+}
+
+#[tokio::test]
+async fn crypto_rsa_pss_sign_verify() {
```
Comment (QuiiBz) [inline crates/runtime/tests/crypto.rs:614]: I've finally found the reason why this test fails:

First, it fails due to a timeout implemented in the runtime. I increased it locally to debug using `IsolateOptions::tick_timeout` & `IsolateOptions::total_timeout`.

Digging a bit more, I found that `RsaPrivateKey::new_with_exp` took +3s to run, and [looking at the readme](https://github.com/RustCrypto/RSA#example) it's somehow expected. Using an `opt-level` of 3 and the method now takes 150ms:

Screenshot 2023-05-14 at 10 39 09

File: crates/runtime_crypto/Cargo.toml (hunk 1 above)
Comment (akitaSummer) [reply crates/runtime_crypto/Cargo.toml]: yes, we can use the latest version.

File: crates/runtime/tests/crypto.rs (hunk 2 above)
Comment (akitaSummer) [reply crates/runtime/tests/crypto.rs:614]: done

File: crates/runtime_crypto/src/methods/generate_key.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                None => return Err(anyhow!("Algorithm modulusLength not found")),
+            };
+
+            let public_exponent_key = v8_string(scope, "publicExponent");
+            let public_exponent = match algorithm.get(scope, public_exponent_key.into()) {
+                Some(public_exponent) => extract_v8_uint8array(public_exponent)?,
+                None => return Err(anyhow!("Algorithm publicExponent not found")),
+            };
+
+            // let sha = get_sha_hash_from_v8_obj(scope, algorithm)?;
```
Comment (QuiiBz) [inline crates/runtime_crypto/src/methods/generate_key.rs:57]: Could we add the sha here? It seems required https://w3c.github.io/webcrypto/#dfn-RsaHashedKeyGenParams

File: crates/runtime/tests/crypto.rs [hunk 4]
```diff
@@ -548,3 +567,131 @@ async fn crypto_derive_key() {
...
+    utils::setup();
+    let (send, receiver) = utils::create_isolate(IsolateOptions::new(
+        "export async function handler() {
+        const keypair_1 = await crypto.subtle.generateKey(
+            {
+                name: 'RSA-PSS',
+                modulusLength: 2048,
```
Comment (QuiiBz) [inline crates/runtime/tests/crypto.rs:621]: Could we use a smaller `modulusLength` for both `crypto_rsa_pss_sign_verify` & `crypto_rsa_ssa_sign_verify`? It's still timeouting randomly on the CI, using e.g 1024 always takes <100ms on my machine

**Review by QuiiBz — CHANGES_REQUESTED** [review]

File: crates/runtime/tests/crypto.rs (hunk 4 above)
Comment (akitaSummer) [reply crates/runtime/tests/crypto.rs]: done

Comment (akitaSummer) [reply crates/runtime/tests/crypto.rs]: However, I noticed that when running the crypto_encrypt_aes_gcm test, an error is thrown stating "Uncaught Value is not of type 'Uint8Array'".

File: crates/runtime_crypto/src/methods/generate_key.rs (hunk 3 above)
Comment (akitaSummer) [reply crates/runtime_crypto/src/methods/generate_key.rs:57]: image
The problem is that when calling sign or verify, the parameter key_value: Vec<u8> is used, so I cannot know which hash RSA is using.

Comment (QuiiBz) [reply crates/runtime_crypto/src/methods/generate_key.rs:57]: Hum got it, we can tackle this later and always use a sha256 for RSA

**Review by QuiiBz — APPROVED** [review]
LGTM, thank you so much!

**Pull request merged**